### ISSUE TITLE
job-exec: replace fork/waitpid with a mechanism which allows "rediscovery" of broker subprocesses

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#3815]
## Problem

Currently, job shells are launched in bulk via libsubprocess from the job-exec module on rank 0 of the system instance. If a broker or subtree needs to be restarted, the flux-shell processes will be signaled (or an attempt to signal them will be made, this may not work for multi-user jobs), and on restart all knowledge of the subprocesses will be lost. Additionally, the job-exec module will receive no further RPC responses in the libsubprocess protocol and will assume the shell processes are still running, or at best will receive an error response and assume the shells are dead and will terminate the job.

As part of [L2 Resiliency](https://github.com/flux-framework/flux-core/wiki/System-Instance-L2-Resiliency) a broker that is restarted will need the ability to leave these launched flux-shell processes running after shutdown, as well as a method to discover any actively running job shells on restart.

A restarted broker will additionally need to be able to do the equivalent of `waitpid()` on processes for which it is no longer the parent.

This is one step in allowing jobs to survive a broker restart.

## Experiment: run job shells as transient service units in a user systemd instance (session manager)

A potential solution, specific to systems running systemd, is to leverage systemd to manage containment and execution of job shells. Thus, the systemd instance will become the parent of all job shells. The flux would use the DBus API to request launch of jobs, and would also use DBus to get asynchronously notified of status changes in the transient unit (e.g. PID, exit status, etc). The broker can also set the `RemainAfterExit` property on the unit, so that information about a job shell that exits is not lost if it occurs while a broker is disconnected or restarting.

The first requirement here is that a user systemd instance for the user `flux` is started when the `flux.service` is started. This is not as easy as it seems, ideally we'd be able to add:

```
Wants: user@flux.service
```

Unfortunately, the correct service name for user systemd instance is `user@UID.service`, and there does not seem to be a way to look up a UID in a unit file, so we have to settle for starting systemd via `ExecStartPre`:
```
# Ensure the user systemd instance for user 'flux' is started.
#  The system instance will use this to run job shells as
#  transient units.
ExecStartPre=bash -c 'systemctl start user@$(id -u flux).service'
```

We also need to ensure that the Flux instance will have access to the DBus connection. Again, since the `flux` userid can't be determined in the unit file, we have to set the requisite environment variables on the `ExecStart` command line:

```
ExecStart=bash -c '\
  XDG_RUNTIME_DIR=/run/user/$UID \
  DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/$UID/bus \
  @X_BINDIR@/flux broker \
  --config-path=@X_SYSCONFDIR@/flux/system/conf.d \
  -Stbon.fanout=256 \
  -Srundir=@X_RUNSTATEDIR@/flux \
  -Slocal-uri=local://@X_RUNSTATEDIR@/flux/local \
  -Slog-stderr-level=6 \
  -Slog-stderr-mode=local \
  -Scontent.backing-path=@X_LOCALSTATEDIR@/lib/flux/content.sqlite \
  -Sbroker.rc2_none \
  -Sbroker.quorum=0 \
'
```

The next part of the experiment was to replace the exec of `flux-imp flux-shell JOBID` with `systemd-run --user [OPTIONS] flux-imp flux-shell JOBID`. `systemd-run` is a user level utility that runs transient units. It uses the `libsd-bus` API, so will be a good reference when implementing DBus calls in Flux directly.

In the experiment, I ended up with the following change (apologies for lack of error checking):

```diff
diff --git a/src/modules/job-exec/exec.c b/src/modules/job-exec/exec.c
index 7bb325f05..1aa49d86a 100644
--- a/src/modules/job-exec/exec.c
+++ b/src/modules/job-exec/exec.c
@@ -195,6 +195,7 @@ static int exec_init (struct jobinfo *job)
     struct exec_conf *conf = NULL;
     struct bulk_exec *exec = NULL;
     const struct idset *ranks = NULL;
+    char idstr[64];
 
     if (job->multiuser && !flux_imp_path) {
         flux_log (job->h,
@@ -228,6 +229,18 @@ static int exec_init (struct jobinfo *job)
         flux_log_error (job->h, "exec_init: flux_cmd_setenvf");
         goto err;
     }
+    if (flux_job_id_encode (job->id, "f58", idstr, sizeof (idstr)) < 0) {
+        flux_log_error (job->h, "exec_init: flux_cmd_setenvf");
+        goto err;
+    }
+    flux_cmd_argv_append (cmd, "systemd-run");
+    flux_cmd_argv_append (cmd, "--user");
+    //flux_cmd_argv_append (cmd, "--remain-after-exit");
+    flux_cmd_argv_append (cmd, "--pipe");
+    flux_cmd_argv_append (cmd, "--collect"); // For now
+    flux_cmd_argv_append (cmd, "--property=CPUAccounting=1");
+    flux_cmd_argv_appendf (cmd, "--setenv=FLUX_KVS_NAMESPACE=%s", job->ns);
+    flux_cmd_argv_appendf (cmd, "--unit=%s", idstr);
     if (job->multiuser) {
         if (flux_cmd_argv_append (cmd, flux_imp_path) < 0
             || flux_cmd_argv_append (cmd, "exec") < 0) {

```

In the final result, `systemd-run` is a child of the broker, but the `flux-shell` and its descendents are all children of the `flux` user's systemd instance. E.g.:

```
[flux@cb26203474f0 ~]$ systemctl status --user 
● cb26203474f0
    State: degraded
     Jobs: 0 queued
   Failed: 1 units
    Since: Wed 2021-08-04 20:57:02 UTC; 1 day 2h ago
   CGroup: /system.slice/snap.docker.dockerd.service/user.slice/user-1000.slice/user@1000.service
           ├─fDNbAhLJRM.service
           │ ├─532 /usr/libexec/flux/flux-shell 1584425347317760
           │ └─534 /usr/bin/coreutils --coreutils-prog-shebang=sleep /usr/bin/sleep 300
           ├─fDP3q6MBCj.service
           │ ├─622 /usr/libexec/flux/flux-shell 1585440335331328
           │ ├─624 broker
           │ └─667 /bin/bash
           ├─init.scope
           │ ├─46 /usr/lib/systemd/systemd --user
           │ └─47 (sd-
           └─dbus.service
             └─149 /usr/bin/dbus-daemon --session --address=systemd: --nofork --nopidfile --systemd-activation --syslog-only
```

Note that two jobs are running in the system instance, one running `sleep 300` and the other is a batch job running a flux-broker.

In this experiment, jobs behave as expected since we have `systemd-run` as the proxy for the job shell. When the transient unit exits, `systemd-run` exits with the correct status, and when a request to signal a job occurs, `systemd-run` is signaled which then presumably forwards the signal to the unit processes.

However, this only demonstrates that process management can be foisted off the the user systemd instance. In order to get the features we need, we'll have to get rid of `systemd-run` and use the DBus API directly.

## Further work

To proceed any further, I think we'll need a replacement for asynchronous subprocess management using the DBus API.

This can then be used as an alternative _in the system instance only_. To allow running jobs to be recovered.

A design something like this might work:

 * When launching jobs, either a new flag to libsubprocess, or a job-exec specific RPC requests that job shells be launched as a systemd transient service unit. (Since we'd be "reaping" children via dbus, the caveat about only one libev reactor in a process handling children does not apply)
 * When a broker restarts, it connects to the systemd dbus and lists active units. Any active units that match the running job pattern (maybe we could choose flux-<jobid>.service instead of what I've done above) are "discovered" as currently running jobs/subprocesses.
 * Similarly any inactive units are discovered and an "exit" event queued for them
 * Any "active" jobs are sent upstream

In this way, the rank 0 `job-exec` module could reacquire the state of running subprocesses. (yeah, there's a lot of missing detail for now)

## Next steps

To proceed any further on the prototype, we'll need to demonstrate that the [sd-bus api of systemd](http://0pointer.net/blog/the-new-sd-bus-api-of-systemd.html) can be used with the flux reactor. This API looks a bit painful, but we can use the source of [systemd-run](https://github.com/systemd/systemd/blob/main/src/run/run.c) as a guide.

It also looks like there is a way to process sd-bus messages asynchronously with [sd_bus_get_fd(3)](https://www.freedesktop.org/software/systemd/man/sd_bus_get_fd.html#) [sd_bus_process(3)](https://www.freedesktop.org/software/systemd/man/sd_bus_process.html#)

We can then replace fork/exec/waitpid() with an sd-bus implementation, perhaps in libsubproces (to be determined) and demonstrate that this works.

**Comment by garlick** [comment]
Great work (and fast too!)

**Comment by grondo** [comment]
I forgot to mention above some other benefits of using systemd for launching job shells in the system instance:

 * systemd runs each transient unit in its own cgroup
 * transient units support many systemd unit properties, so we can support containment easily using, e.g. properties described in [systemd.resource-control](https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html)
 * could potentially take advantage of [systemctl freeze](https://www.freedesktop.org/software/systemd/man/systemctl.html#freeze%20PATTERN%E2%80%A6) to "freeze" jobs (needs more research though, didn't work in my testing, perhaps because centos8 isn't using the unified cgroup hierarchy)

Overall, this approach may save us some work up front, at the expense reduced portability (systemd will be required for the multi-user system instance.)

**Comment by grondo** [comment]
> transient units support many systemd unit properties, so we can support containment easily using, e.g. properties described in systemd.resource-control

This might imply we want to use this style of launching from `job-exec` only and _not_ enhance the libsubprocess code...

**Comment by chu11** [comment]
Just thought I'd post some notes on experimenting w/ the sd-bus API.  The sd-bus API is not the easiest API to learn, as it's not so widely used as to have plenty of examples / stackoverflow answers online.  So a lot of trial and error and searching online have to go on.

Some links of note

DBUS spec - https://dbus.freedesktop.org/doc/dbus-specification.html
org.freedesktop.systemd1 manpage - https://www.freedesktop.org/software/systemd/man/org.freedesktop.systemd1.html
stackoverflow useful answer - https://stackoverflow.com/questions/61940461/how-to-get-the-state-of-a-service-with-sd-bus
stackoverflow useful answer - https://unix.stackexchange.com/questions/527283/watching-sd-bus-properties

I missed around with the code on this branch:

https://github.com/chu11/systemd/tree/systemd_run_min

The branch also has three files `my-run`, `my-wait`, and `my-cleanup` which are tiny files that run, wait, and cleanup a task running under systemd.  I hope that these files will emulate what we ultimately will put into `flux-core`.  There are notes in `my-notes.txt` in that branch as well.  Here's a sample of what they do (removed some extraneous debug output and re-formated for this post)

```
$ ./my-run --unit=test$RANDOM --remain-after-exit /bin/sh -c "/bin/sleep 60; /bin/false"
unit name service is test8670.service
Running as unit: test8670.service

(lots of debugging, note failed state at the end and exit status is 1)
$ ./my-wait --unit=test8670
unit name service is /org/freedesktop/systemd1/unit/test8670_2eservice
initial active state = active
initial load state = loaded
start wait time = 1630384226
time passed = 0
revents = 1, n = 1
start wait time = 1630384226
time passed = 49
revents = 1, n = 1
new result = exit-code
exit timestamp= 1630384275547615
exit status = 1
new active state = failed

(call my-wait again, we catch if its already done)
$ ./my-wait --unit=test8670
unit name service is /org/freedesktop/systemd1/unit/test8670_2eservice
initial active state = failed
initial load state = loaded
unit test8670 is already done
exit status = 1
start time = 1630384215538223
Exit time = 1630384275547615

(failed unit is cached by systemd)
$ systemctl list-units --user | grep test
test8670.service                loaded failed failed    running a job test

(internally do equivalent of systemctl reset-failed --user ...)
$ ./my-cleanup --unit=test8670
unit name service path is /org/freedesktop/systemd1/unit/test8670_2eservice
initial active state = failed
unit name service name is test8670.service
reset failed unit test8670

$ systemctl list-units --user | grep test
$
```


some side notes

- the `my-wait` above waits for a task's completion via `poll()`, which hopefully we can use with a fd watcher in flux (you'll note the "time passed" is 49 seconds, about right for waiting for the task's "sleep 60" to complete.)

- there is a small "racy window" in the polling:

  - A) check for a job being done, if it's not done
  - B) poll for it to be done

  there's a small racy window that the above could miss if the unit completes between A & B, then the polling may sit forever for a unit that will never update its state.  Not sure of how to solve at the moment.  It may be worthwhile to have the first poll iteration not be super long (lets say order of ~60 seconds) to protect against this racy window.  Or maybe need to check exit time one more time after setting up the polling.

- the `--remain-after-exit` will leave a successful task in the "active" state, so we have to check for unit being done via exit time.  I hope this is correct / ok.  OR perhaps there's an alternate way to tell systemd to not punt task results, but I have yet to figure this out.

**Comment by chu11** [comment]
Thought I'd post this initial `libsdprocess` header file.  It's not pretty, but thought I'd post this anyways for some initial comments.

```
typedef struct flux_sdprocess flux_sdprocess_t;

typedef void (*flux_sdprocess_f) (flux_sdprocess_t *sdp, void *arg);

/* present assumptions:
 * unitname 1 word, no dashes, periods, etc.
 * command is absolute pathed
 * argv & env NULL terminated
 * fd's < 0 if don't want to use
 */
flux_sdprocess_t *flux_sdprocess_exec (flux_t *h,
                                       const char *unitname,
                                       char **argv,
                                       char **envv,
                                       int stdin_fd,
                                       int stdout_fd,
                                       int stderr_fd);

/* for testing */
flux_sdprocess_t *flux_sdprocess_local_exec (flux_reactor_t *r,
                                             const char *unitname,
                                             char **argv,
                                             char **envv,
                                             int stdin_fd,
                                             int stdout_fd,
                                             int stderr_fd);

int flux_sdprocess_wait (flux_sdprocess_t *sdp);

int flux_sdprocess_completed_callback (flux_sdprocess_t *sdp,
                                       flux_sdprocess_f completed_cb,
                                       void *arg);

bool flux_sdprocess_completed (flux_sdprocess_t *sdp);

/* XXX
 * need unitname variants, i.e.
 * flux_sdprocess_wait_unitname (const char *unitname);
 */

int flux_sdprocess_exit_status (flux_sdprocess_t *sdp);

/* different than destroy, cleanup leftovers in systemd */
int flux_sdprocess_systemd_cleanup (flux_sdprocess_t *sdp);

void flux_sdprocess_destroy (flux_sdprocess_t *sdp);
```

and here's a sample of its usage in my tests

```
    sdp = flux_sdprocess_local_exec (r,
                                     unitname,
                                     cmdv,
                                     NULL,
                                     STDIN_FILENO,
                                     STDOUT_FILENO,
                                     STDERR_FILENO);
    ok (sdp != NULL,
        "flux_sdprocess_local_exec launched process under systemd");

    ret = flux_sdprocess_completed_callback (sdp, count_cb, &count);
    ok (ret == 0,
        "flux_sdprocess_completed_callback success");

    if (!count) {
        ret = flux_reactor_run (r, 0);
        ok (ret == 0,
            "flux_reactor_run success");
    }

    ok (count == 1,
        "callback called once");

    /* OR */
    /* ret = flux_sdprocess_wait (sdp); */
    /* ok (ret == 0, */
    /*     "flux_sdprocess_wait success"); */

    ret = flux_sdprocess_exit_status (sdp);
    ok (ret == 0,
        "flux_sdprocess_exit_status returns correct exit code");

    ret = flux_sdprocess_systemd_cleanup (sdp);
    while (ret < 0 && errno == EBUSY)
        ret = flux_sdprocess_systemd_cleanup (sdp);
    ok (ret == 0,
        "flux_sdprocess_systemd_cleanup success");
```

There's a lot of "polish" that needs to be done / can be done, but some can be punted to another day as we iterate over this more and more.  (e.g. command must be in an absolute path, unitname cannot have special characters, how to deal with stdio better, run under different user/group/working directory, allow caller to set systemd description, better errnos to return, dunno if sending signals is needed, etc.)

But I was trying to think of the steps to get us to an initial "checkpoint" we can commit into flux-core.

- is there some "bare minimum functionality" the initial checkpoint should have?  (maybe need a few getters, like `get_unitname()` and stuff)

- how to test.  I'm testing via `make check` in docker (via script in https://github.com/flux-framework/flux-core/pull/3844).  I could just have the tests automatically pass if systemd and appropriate environment stuff isn't setup.  But at some point we'd like to have `make check` work under a docker image all the time.

- should we add the tweaks in https://github.com/flux-framework/flux-core/issues/3815 to startup flux user service by default?

- need to add systemd-devel into the fluxorama image, not sure the best place to do this.  I lazily just did a yum install in the fluxorama Dockerfile, but I think we could add it to some "parent" container for fluxorama.

- would be nice to add some devel packages into the fluxorama image too (gdb was missing, emacs would be nice)

**Comment by grondo** [comment]
Awesome that you seem to have gotten stdio working! It seems you are getting really close to being able to test this as an alternate "exec method" in an instance of size 1 (since job-exec doesn't yet run on all ranks)

> is there some "bare minimum functionality" the initial checkpoint should have? (maybe need a few getters, like get_unitname() and stuff)

Just thinking out loud, but we could start out perhaps by plugging this in as a new job-exec "implementation" (the existing job-exec module has `testexec` and `bulkexec` implementations). This would allow further testing on a single rank instance without being blocked by a job-exec rewrite.

The exec_implementation methods are described here:
https://github.com/flux-framework/flux-core/blob/6dc5e29d96cba23526a65d59009176cd70bd311e/src/modules/job-exec/job-exec.h#L24-L54

Additionally, the exec implementation should be able to update the job state with the job-exec module in the following places:

 - call `jobinfo_started()` when all job shells have started (in this case just the 1 shell)
 - call `jobinfo_fatal_error()` whenever a fatal error occurs
 - call `jobfino_tasks_complete()` when all job shells exit (again just 1 job shell in this case)
 - call `jobinfo_log_output()` for any output from the job shell (or systemd in this case)
 - write `J` to `stdin` of the job shell (see `exec.c:start_cb()`)

The next step would be to ensure that a process can discover any running jobs left after a restart. The running (or exited/failed) units would need to be associated with a job, so the unit name should perhaps always contain the jobid.

> how to test. I'm testing via make check 

My idea was to put a suite of "system" sharness tests in a subdirectory so those could be invoked under the fluxorama test image. It occurs to me though you can do some low level testing of your `flux_sdprocess_*` interface outside of the fluxorama image if a systemd user session exists for the user running `make check`. We can probably make that happen in CI quite easily. Your unit test should just check for existence of the appropriate systemd facility and "skip all" if it doesn't exist.

Configure should probably also check for systemd-devel and disable generation of this library and its dependent facilities so that we don't require systemd-devel for building flux-core when it is only being used for single-user mode.

I can add the dependencies to the `fluxrm/testenv:centos8` image. We probably want to avoid adding bloat to that image though (it is already huge), so probably best to continue manually installing gdb and emacs after the fact. I tend to develop outside of the docker image and only run build/test inside.
